>[ClickHouse special build check — 8/9 artifact groups are OK (8 of 9 builds are OK)](https://s3.amazonaws.com/clickhouse-builds/42582/85e41a1ecabd9058a90720e2facd48fa3452da51/clickhouse_special_build_check/report.html)

clang-tidy

```
Requested labels: self-hosted, builder
Job defined at: ClickHouse/ClickHouse/.github/workflows/pull_request.yml@refs/pull/42582/merge
Waiting for a runner to pick up this job...
```

>[Integration tests (asan) [1/3] — fail: 1, passed: 666, flaky: 4](https://s3.amazonaws.com/clickhouse-test-reports/42582/85e41a1ecabd9058a90720e2facd48fa3452da51/integration_tests__asan__[1/3].html)

- `test_multiple_disks/test.py::test_freeze`

Failed one time before - https://play.clickhouse.com/play?user=play#c2VsZWN0IAogIGRpc3RpbmN0IHJlcG9ydF91cmwKZnJvbSBjaGVja3MKd2hlcmUKICBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gaW50ZXJ2YWwgNyBkYXkgYW5kCiAgdGVzdF9uYW1lIGxpa2UgJyV0ZXN0X211bHRpcGxlX2Rpc2tzL3Rlc3QucHk6OnRlc3RfZnJlZXplJScgYW5kCiAgdGVzdF9zdGF0dXMgaW4gKCdGQUlMJywgJ0ZMQUtZJywgJ1NLSVBQRUQnKQ==

>[Integration tests (tsan) [2/4] — fail: 4, passed: 524, flaky: 4](https://s3.amazonaws.com/clickhouse-test-reports/42582/85e41a1ecabd9058a90720e2facd48fa3452da51/integration_tests__tsan__[2/4].html)

- `test_backup_restore_on_cluster/test_concurrency.py`

```
E           Code: 159. DB::Exception: Received from 172.16.6.14:9000. DB::Exception: Watching task /clickhouse/task_queue/ddl/query-0000000207 is executing longer than distributed_ddl_task_timeout (=180) seconds. There are 1 unfinished hosts (0 of them are currently active), they are going to execute the query in background. Stack trace:
```

>[Stateless tests (tsan, s3 storage) [1/3] — fail: 1, passed: 1521, skipped: 22](https://s3.amazonaws.com/clickhouse-test-reports/42582/85e41a1ecabd9058a90720e2facd48fa3452da51/stateless_tests__tsan__s3_storage__[1/3].html)

- `02457_datediff_via_unix_epoch` - #42655

--- a/src/Common/MemoryTracker.cpp
+++ b/src/Common/MemoryTracker.cpp
@@ -10,6 +10,7 @@
 #include <Common/ProfileEvents.h>
 #include <Common/thread_local_rng.h>
 #include <Common/OvercommitTracker.h>
+#include <Common/Stopwatch.h>
 #include <Common/logger_useful.h>
 
 #include "config.h"
@@ -86,6 +87,8 @@ inline std::string_view toDescription(OvercommitResult result)
 namespace ProfileEvents
 {
     extern const Event QueryMemoryLimitExceeded;
+    extern const Event MemoryAllocatorPurge;
+    extern const Event MemoryAllocatorPurgeTimeMicroseconds;
 }
 
 using namespace std::chrono_literals;
@@ -229,7 +232,10 @@ void MemoryTracker::allocImpl(Int64 size, bool throw_if_memory_exceeded, MemoryT
         {
             if (free_memory_in_allocator_arenas.exchange(-current_free_memory_in_allocator_arenas) > 0)
             {
+                Stopwatch watch;
                 mallctl("arena." STRINGIFY(MALLCTL_ARENAS_ALL) ".purge", nullptr, nullptr, nullptr, 0);
+                ProfileEvents::increment(ProfileEvents::MemoryAllocatorPurge);
+                ProfileEvents::increment(ProfileEvents::MemoryAllocatorPurgeTimeMicroseconds, watch.elapsedMicroseconds());
             }
         }
 

--- a/src/Common/MemoryTracker.cpp
+++ b/src/Common/MemoryTracker.cpp
@@ -438,7 +438,7 @@ void MemoryTracker::reset()
 
 void MemoryTracker::setRSS(Int64 rss_, Int64 free_memory_in_allocator_arenas_)
 {
-    Int64 new_amount = rss_; // - free_memory_in_allocator_arenas_;
+    Int64 new_amount = rss_;
     total_memory_tracker.amount.store(new_amount, std::memory_order_relaxed);
     free_memory_in_allocator_arenas.store(free_memory_in_allocator_arenas_, std::memory_order_relaxed);
 

--- a/src/Common/ProfileEvents.cpp
+++ b/src/Common/ProfileEvents.cpp
@@ -229,6 +229,8 @@ The server successfully detected this situation and will download merged part fr
     M(UserTimeMicroseconds, "Total time spent in processing (queries and other tasks) threads executing CPU instructions in user space. This include time CPU pipeline was stalled due to cache misses, branch mispredictions, hyper-threading, etc.") \
     M(SystemTimeMicroseconds, "Total time spent in processing (queries and other tasks) threads executing CPU instructions in OS kernel space. This include time CPU pipeline was stalled due to cache misses, branch mispredictions, hyper-threading, etc.") \
     M(MemoryOvercommitWaitTimeMicroseconds, "Total time spent in waiting for memory to be freed in OvercommitTracker.") \
+    M(MemoryAllocatorPurge, "Total number of times memory allocator purge was requested") \
+    M(MemoryAllocatorPurgeTimeMicroseconds, "Total number of times memory allocator purge was requested") \
     M(SoftPageFaults, "The number of soft page faults in query execution threads. Soft page fault usually means a miss in the memory allocator cache which required a new memory mapping from the OS and subsequent allocation of a page of physical memory.") \
     M(HardPageFaults, "The number of hard page faults in query execution threads. High values indicate either that you forgot to turn off swap on your server, or eviction of memory pages of the ClickHouse binary during very high memory pressure, or successful usage of the 'mmap' read method for the tables data.") \
     \

--- a/src/Interpreters/AsynchronousMetrics.cpp
+++ b/src/Interpreters/AsynchronousMetrics.cpp
@@ -703,12 +703,18 @@ void AsynchronousMetrics::update(TimePoint update_time)
             Int64 free_memory_in_allocator_arenas = 0;
 
 #if USE_JEMALLOC
-            /// This is a memory which is kept by allocator.
-            /// Will subsract it from RSS to decrease memory drift.
+            /// According to jemalloc man, pdirty is:
+            ///
+            ///     Number of pages within unused extents that are potentially
+            ///     dirty, and for which madvise() or similar has not been called.
+            ///
+            /// So they will be subtracted from RSS to make accounting more
+            /// accurate, since those pages are not really RSS but a memory
+            /// that can be used at anytime via jemalloc.
             free_memory_in_allocator_arenas = je_malloc_pdirty * getPageSize();
 #endif
 
-            Int64 difference = rss - free_memory_in_allocator_arenas - amount;
+            Int64 difference = rss - amount;
 
             /// Log only if difference is high. This is for convenience. The threshold is arbitrary.
             if (difference >= 1048576 || difference <= -1048576)

--- a/src/Interpreters/AsynchronousMetrics.cpp
+++ b/src/Interpreters/AsynchronousMetrics.cpp
@@ -713,9 +713,10 @@ void AsynchronousMetrics::update(TimePoint update_time)
             /// Log only if difference is high. This is for convenience. The threshold is arbitrary.
             if (difference >= 1048576 || difference <= -1048576)
                 LOG_TRACE(log,
-                    "MemoryTracking: was {}, peak {}, will set to {} (RSS), difference: {}",
+                    "MemoryTracking: was {}, peak {}, free memory in arenas {}, will set to {} (RSS), difference: {}",
                     ReadableSize(amount),
                     ReadableSize(peak),
+                    ReadableSize(free_memory_in_allocator_arenas),
                     ReadableSize(rss),
                     ReadableSize(difference));
 

